go/runtime/bundle/discovery.go

```diff
@@ -0,0 +1,452 @@
+	var hashes []hash.Hash
+	for _, hash := range manifestHashes {
+		if d.registry.HasBundle(hash) {
+			continue
+		}
+		hashes = append(hashes, hash)
+	}
+
+	// Update the queue with the new hashes.
+	if len(hashes) == 0 {
+		delete(d.manifestHashes, runtimeID)
+		return
+	}
+	d.manifestHashes[runtimeID] = hashes
+
+	// Trigger immediate discovery or download of new bundles.
+	select {
+	case d.discoverCh <- struct{}{}:
+	default:
+	}
+}
+
+// Download tries to download bundles in the queue.
+func (d *Discovery) Download() {
+	d.mu.RLock()
+	runtimeIDs := slices.Collect(maps.Keys(d.manifestHashes))
+	d.mu.RUnlock()
+
+	for _, runtimeID := range runtimeIDs {
+		d.downloadBundles(runtimeID)
+	}
+}
+
+func (d *Discovery) downloadBundles(runtimeID common.Namespace) {
+	// Try to download queued bundles.
+	d.mu.RLock()
+	hashes := d.manifestHashes[runtimeID]
+	d.mu.RUnlock()
+
+	downloaded := make(map[hash.Hash]struct{})
+	for _, hash := range hashes {
+		if err := d.downloadBundle(runtimeID, hash); err != nil {
+			d.logger.Error("failed to download bundle",
+				"err", err,
+				"runtime_id", runtimeID,
+				"manifest_hash", hash.Hex(),
+			)
+			continue
+		}
+		downloaded[hash] = struct{}{}
+	}
+
+	// Remove downloaded bundles from the queue.
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	var pending []hash.Hash
+	for _, hash := range d.manifestHashes[runtimeID] {
+		if _, ok := downloaded[hash]; ok {
+			continue
+		}
+		pending = append(pending, hash)
+	}
+	if len(pending) == 0 {
+		delete(d.manifestHashes, runtimeID)
+		return
+	}
+	d.manifestHashes[runtimeID] = pending
+}
+
+func (d *Discovery) downloadBundle(runtimeID common.Namespace, manifestHash hash.Hash) error {
+	var errs error
+
+	for _, baseURLs := range [][]string{d.runtimeBaseURLs[runtimeID], d.globalBaseURLs} {
+		for _, baseURL := range baseURLs {
+			if err := d.tryDownloadBundle(runtimeID, manifestHash, baseURL); err != nil {
+				errs = errors.Join(errs, err)
+				continue
+			}
+
+			return nil
+		}
+	}
+
+	return errs
+}
+
+func (d *Discovery) tryDownloadBundle(runtimeID common.Namespace, manifestHash hash.Hash, baseURL string) error {
+	filename := fmt.Sprintf("%s%s", manifestHash.Hex(), FileExtension)
+
+	d.logger.Debug("downloading bundle",
+		"runtime_id", runtimeID,
+		"base_url", baseURL,
+		"filename", filename,
+	)
+
+	url, err := url.JoinPath(baseURL, filename)
+	if err != nil {
+		d.logger.Error("failed to construct URL",
+			"err", err,
+			"base_url", baseURL,
+			"filename", filename,
+		)
+		return fmt.Errorf("failed to construct URL: %w", err)
+	}
+
+	src, err := d.fetchBundle(url)
+	if err != nil {
+		d.logger.Error("failed to download bundle",
+			"err", err,
+			"url", url,
+		)
+		return fmt.Errorf("failed to download bundle: %w", err)
+	}
+	defer os.Remove(src)
+
+	d.logger.Info("bundle downloaded",
+		"runtime_id", runtimeID,
+		"base_url", baseURL,
+		"filename", filename,
+	)
+
+	if err := d.registry.AddBundle(src, manifestHash); err != nil {
```

Comment:
It is a bit scary that `AddBundle` needs to "Open" the (untrusted) bundle before verifying the manifest hash.  I am not sure, but the current implementation of `Open` might assume that bundle came from a (somewhat) trusted source, which might not be the case in future. We should ensure that `Open` is safe to use with malicious inputs. 

---

Somehwat related, I remember we talked about https://github.com/oasisprotocol/oasis-core/commit/38edded4cad5cabbe9a78953880448ba540d0d02 . Now that I have more context, shouldn't that actually be the hash of the entire bundle (not just the manifest)? This would prevent this problem and allow us to verify the downloaded file before opening it.

Reply:
Validation is actually hidden in the `Open` function, so the bundle doesn't need to be trusted.

We can refactor this to use bundle checksum instead (maybe in another PR), as that might be more intuitive. However, the current approach allows us to remove the bundles entirely as one could just publish the manifest (e.g. on chain) and the users should then fetch binaries from untrusted sources and verify their checksums.

@kostko What are your thoughts?

Reply:
The idea for using the manifest hash was that one would not need to always download all of the bundle artifacts as some may be shared between bundles.

We should probably add a separate `Open` variant (or have the existing one take options), one that takes a manifest hash. Obviously the zip archive parser should be well-fuzzed (which the current implementation is) and since the manifest _must_ be the first file in the archive, we can just verify the manifest hash before continuing to load or parse anything else. I would opt for this variant.

Reply:
BTW, we should do some bundle fuzzing, just to be sure these things are somewhat robust. Will open an issue.

Reply:
#5969

Reply:
@kostko I think in this context it would be worth adding some additional doc of the current and intendend status of bundles.

Here is what I have in mind:
* right now we have legacy and non-legacy bundles (components)
* furthermore we will have two ways of adding bundles (via the `paths` or configuring them to download).
* finally, there is this things with versions that is not 100% clear yet (what is allowed to be reused, what if you get same version but different content, exact bundle validation).

I think it may be worth aggregating requirements/invariant somewhere together with legacy context, as in addition to make it more accessible it would also ease us in finding problematic corner cases/interleaving. Maybe a `README.md` under `go/runtime/bundle`? We can later move this to official docs once those features are released?

Hope I am not complicating, we could also offline :).

Reply:
I think that runtime bundles don't need to be compatible with all previous versions, so I expect we will remove all these legacy fields once we are done refactoring and all runtimes upgrade to new bundles. So I'm not sure if we need a readme file.